src/main/java/com/google/devtools/build/lib/analysis/config/ExecutionTransitionFactory.java

```diff
@@ -0,0 +1,64 @@
+package com.google.devtools.build.lib.analysis.config;
+
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.lib.analysis.PlatformOptions;
+import com.google.devtools.build.lib.analysis.config.transitions.PatchTransition;
+import com.google.devtools.build.lib.analysis.config.transitions.TransitionFactory;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.packages.AttributeTransitionData;
+import javax.annotation.Nullable;
+
+public class ExecutionTransitionFactory implements TransitionFactory<AttributeTransitionData> {
+
+  @Override
+  public PatchTransition create(AttributeTransitionData data) {
+    return new ExecutionTransition(data.executionPlatform());
+  }
+
+  @Override
+  public boolean isHost() {
+    return false;
+  }
+
+  private static class ExecutionTransition implements PatchTransition {
+    @Nullable private final Label executionPlatform;
+
+    public ExecutionTransition(@Nullable Label executionPlatform) {
+      this.executionPlatform = executionPlatform;
+    }
+
+    @Override
+    public String getName() {
+      return "exec";
+    }
+
+    @Override
+    public boolean isHostTransition() {
+      return false;
+    }
+
+    @Override
+    public BuildOptions patch(BuildOptions options) {
+      if (executionPlatform == null) {
+        // No execution platform is known, so don't change anything.
+        return options;
+      }
+
+      // Start by converting to host options.
+      BuildOptions execConfiguration = options.createHostOptions();
+
+      // Then unset isHost, if CoreOptions is available.
```

Comment:
I'm not opposed to this check, but are you imagining an actual scenario when CoreOptions isn't available? In practice I think it's always going to remain "universal", even with trimming. In my mind anything worthy of trimming would eventually be migrated out to its own dedicated fragment.

Reply:
You are correct that in any case where it's not available things have gone very wrong. I've changed it to a Precondition check just to capture the assertion that it must be available.